### PR TITLE
Dbixdpnc

### PR DESCRIPTION
**Pull request by psladek**
action #5852 - remove DBIx::Class::Schema::Config

loading database.ini via Config::IniFiles,
moving connect_db() moved from Utils.pm to Schema.pm

### CODE REVIEW

**Comment by coolo** [comment]
the test suite protests heavily

**Comment by psladek** [comment]
tests access to config file should be fixed now

File: lib/OpenQA/Schema/Schema.pm [hunk 1]
```diff
@@ -24,12 +25,22 @@ use FindBin qw($Bin);
...
-__PACKAGE__->config_paths(\@paths);
+
+sub connect_db {
+    my $mode = shift || $ENV{OPENQA_DATABASE} || 'production';
+    CORE::state $schema;
+    unless ($schema) {
+        my %ini;
+        #        print("ENV{OPENQA_DATABASE_CONFIG}=$ENV{OPENQA_DATABASE_CONFIG}\n");
+        tie %ini, 'Config::IniFiles', ( -file => $ENV{OPENQA_DATABASE_CONFIG} || $configname );
+        $schema=__PACKAGE__->connect($ini{$mode});
```
Comment (coolo) [inline lib/OpenQA/Schema/Schema.pm]: as I just documented the use of OPENQA_CONFIG (see #171), why did you change?

Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: oh, I see, sorry, did not noted your change before pull request. I did change it because from the code I had an impression that OPENQA_CONFIG is only for openqa.ini. It was also originaly done because of the test scripts, which, if in deeper level, could not find the database.ini file ../lib (../../lib was needed). Since in the OpenQA::Test::Case was already a OPENQA_CLIENT_CONFIG variable, it seemed natural to have another independent path for database.ini, in something like OPENQA_DATABASE_CONFIG.

If you think it is better as it was described in the documentation, I can change it to derive from OPENQA_CONFIG, but I still think, since we have already independent OPENQA_CLIENT_CONFIG, the database.ini may be justified to have its own env variable.

**Comment by psladek** [comment]
Also, what do you think about this solution:
OPENQA_CONFIG is set, OPENQA_DATABASE_CONFIG unset :  use path from OPENQA_CONFIG
OPENQA_DATABASE_CONFIG set :  use path from OPENQA_DATABASE_CONFIG
?

File: lib/OpenQA/Schema/Schema.pm (hunk 1 above)
Comment (coolo) [inline lib/OpenQA/Schema/Schema.pm]: I would rather drop the OPENQA_CLIENT_CONFIG than having a variable per file

**Comment by coolo** [comment]
whatever you do: document it. and as you document it, you will notice how silly it is :)

File: lib/OpenQA/Schema/Schema.pm (hunk 1 above)
Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: That's also possibility, but since it will restrict all ini files in one directory (assuming same logic for workers.ini as well), we will need to copy/link openqa.ini and database.ini into t/data directory, and also move lib/openqa.ini and lib/database.ini into etc/openqa. Is that ok to do so?

**Comment by psladek** [comment]
I don't really think so, but in any case, I think I'd rather spare myself the effort :)

File: lib/OpenQA/Schema/Schema.pm (hunk 1 above)
Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: also this will force all configs except openqa.ini to have always same names, maybe we should make this so also for openqa.ini and put only the dirname into the OPENQA_CONFIG ?

Comment (aaannz) [inline lib/OpenQA/Schema/Schema.pm]: Btw. workers.ini is staged for deprecation once somebody figures out the way how to do it :) (see action #4654).

ini files shuffling is done in spec file, but I'm not against moving them to etc/openqa since client.ini is already there. Don't forget to update Makefile then. Why copy to t/data?

I didn't get the second comment. Did you mean it will force to same dirname? I'm not against using only dirname in OPENQA_CONFIG (or both and determine at runtime).

Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: copying, or more likely linking to t/data is needed because of t/data/client.conf. And since all cfg files have to have same directory path now, they must be present there too.

Second comment: I did not say same dirname, but same config name. While you can specify whole pathname (dir+file name) in OPENQA_CONFIG for openqa.ini, the other files have fixed names (defaults, and same directory as OPENQA_CONFIG). So I think it would be better to have all configs on same standing. 
Deciding at runtime seems possible, but not so nice, since one have to check whether OPENQA_CONFIG is file or directory everytime any config file needs to be loaded. On the other hand it would make it backward compatible... But then it also dangerously resemble the idea of combining OPENQA_CONFIG and OPENQA_DATABASE_CONFIG, not sure what coolo would think about that, concerning documentation and such... :)

Comment (aaannz) [inline lib/OpenQA/Schema/Schema.pm]: >copying, or more likely linking to t/data is needed because of t/data/client.conf. And since all cfg files have to have same directory path now, they must be present there too.

Yeah, but is it needed? I don't think they are used anywhere in tests.

> Second comment: I did not say same dirname, but same config name. While you can specify whole pathname (dir+file name) in OPENQA_CONFIG for openqa.ini, the other files have fixed names (defaults, and same directory as OPENQA_CONFIG). So I think it would be better to have all configs on same standing. 

Aha, I understood that as a same name (ie. everything called openqa.ini) which didn't make sense to me. Fixed name it is then.

> Deciding at runtime seems possible, but not so nice, since one have to check whether OPENQA_CONFIG is file or directory everytime any config file needs to be loaded.

AFAIK they are loaded once on startup, I don't see where the problem is.

Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: <blockquote>
Yeah, but is it needed? I don't think they are used anywhere in tests.
</blockquote>
yes, it is needed, for example t/api/06-products.t needs database.ini

<blockquote>
AFAIK they are loaded once on startup, I don't see where the problem is.
</blockquote>
It could of course be done, but I think it is still clearer, in code, documentation and usage to have only one meaning for OPENQA_CONFIG. The only problem is backward compatibility..

Comment (aaannz) [inline lib/OpenQA/Schema/Schema.pm]: >
> Yeah, but is it needed? I don't think they are used anywhere in tests.
>
> yes, it is needed, for example t/api/06-products.t needs database.ini
>
> Ah right, by the very same thing you are trying to get rid of I suppose
(Schema::Config). I wasn't paying attention and it is hidden deep. :)
I expected tests to be self contained and if needed, database.ini would be
already there.

Comment (psladek) [inline lib/OpenQA/Schema/Schema.pm]: Not sure what you mean by they would already be there, I think simply linking the ini files from etc/openqa (where lib/database.ini and lib/openq.ini would be moved too) would make tests work as they do now, which means to also test the project's ini files validity. If later some specific test needs special ini file(s) just for itself, it can have them in some another t/data like directory (again linking the ones that need not to be specific).

File: lib/OpenQA/Client.pm [hunk 2]
```diff
@@ -37,7 +36,10 @@ sub new {
     }
 
     if ($args{api}) {
-        for my $file ($ENV{OPENQA_CLIENT_CONFIG}||undef, glob('~/.config/openqa/client.conf'), '/etc/openqa/client.conf') {
+        my @cfgpaths=('~/.config/openqa', '/etc/openqa');
+        @cfgpaths=($ENV{OPENQA_CONFIG},@cfgpaths) if defined $ENV{OPENQA_CONFIG};
```
Comment (lnussel) [inline lib/OpenQA/Client.pm]: you can use unshift here

File: lib/OpenQA/Client.pm [hunk 3]
```diff
@@ -37,7 +36,10 @@ sub new {
     }
 
     if ($args{api}) {
-        for my $file ($ENV{OPENQA_CLIENT_CONFIG}||undef, glob('~/.config/openqa/client.conf'), '/etc/openqa/client.conf') {
+        my @cfgpaths=('~/.config/openqa', '/etc/openqa');
```
Comment (lnussel) [inline lib/OpenQA/Client.pm]: missing glob, ~ would not be expanded without

File: lib/OpenQA/Client.pm (hunk 2 above)
Comment (psladek) [inline lib/OpenQA/Client.pm]: I know. I am more used to this form though, and in general find it a bit more readable and self explaining.
Can change it if you insist though...

File: lib/OpenQA/Client.pm (hunk 3 above)
Comment (psladek) [inline lib/OpenQA/Client.pm]: thanks, good point, did not realize it is needed for that

**Comment by coolo** [comment]
fine with me now

**Comment by coolo** [comment]
I tried to rebase this, but it's too long history, so merging as is
